packages/core-data/src/reducer.js

```diff
@@ -322,32 +323,26 @@ export function undo( state = UNDO_INITIAL_STATE, action ) {
 				return nextState;
 			}
 
-			let nextState;
-			if ( state.length === 0 ) {
-				// Create an initial entry so that we can undo to it.
-				nextState = [
-					{
-						kind: action.meta.undo.kind,
-						name: action.meta.undo.name,
-						recordId: action.meta.undo.recordId,
-						edits: { ...state.flattenedUndo, ...action.meta.undo.edits },
-					},
-				];
-				nextState.offset = 0;
-				return nextState;
-			}
-
 			// Clear potential redos, because this only supports linear history.
-			nextState = state.slice( 0, state.offset || undefined );
+			const nextState = state.slice( 0, state.offset || undefined );
 			nextState.offset = 0;
-
+			nextState.pop();
 			nextState.push( {
-				kind: action.kind,
-				name: action.name,
-				recordId: action.recordId,
-				edits: { ...action.edits, ...state.flattenedUndo },
+				kind: action.meta.undo.kind,
+				name: action.meta.undo.name,
+				recordId: action.meta.undo.recordId,
+				edits: { ...state.flattenedUndo, ...action.meta.undo.edits },
 			} );
-
+			const comparisonUndoEdits = Object.values( action.meta.undo.edits ).filter( ( edit ) => typeof edit !== 'function' );
```

Comment:
What is this checking for? Would be great to inline comment.

Reply:
That the last two edits are not the same.

Reply:
I'm more wondering what the filter is for. :)

Reply:
When an edit is a function it's an optimization to avoid running some expensive operation. We can't rely on the function references being the same so we opt out of comparing them here.

Reply:
That could use an inline comment.

Reply:
I'm working on it with another fix.

Reply:
#17452